pkg/changelog/changelog.go

```diff
@@ -6,219 +6,229 @@ import (
-	left, right := 0, len(clv.changes)
+// GetAll retrieves all values for the specified `MemberKind`, from the newest to the oldest.
+func (e *Entries[T]) GetAll(kind MemberKind) []T {
+	if e.noEntries(kind) {
+		return nil
+	}
 
-	for left < right {
-		middle := (left + right) / 2
-		if clv.changes[middle].timestamp.After(target) {
-			right = middle
-		} else {
-			left = middle + 1
-		}
+	values := make([]T, 0, e.kindList[kind].size)
+	current := e.kindList[kind].tail
+	for current != nil {
+		values = append(values, current.value)
+		current = current.prev
 	}
 
-	return left
+	return values
 }
 
-// enforceSizeBoundary ensures that the size of the inner array doesn't exceed the limit.
-// It applies two methods to reduce the log size to the maximum allowed:
-// 1. Unite duplicate values that are trailing one another, removing the oldest of the pair.
-// 2. Remove the oldest logs as they are likely less important.
-
-func (clv *Changelog[T]) enforceSizeBoundary() {
-	if len(clv.changes) <= clv.maxSize {
-		// Nothing to do
-		return
+// Count returns the number of entries recorded for the specified `MemberKind`.
+func (e *Entries[T]) Count(kind MemberKind) int {
+	if e.noEntries(kind) {
+		return 0
 	}
 
-	boundaryDiff := len(clv.changes) - clv.maxSize
-	changed := false
-
-	// Compact the slice in place
-	writeIdx := 0
-	for readIdx := 0; readIdx < len(clv.changes); readIdx++ {
-		nextIdx := readIdx + 1
-		if nextIdx < len(clv.changes) &&
-			clv.changes[nextIdx].value == clv.changes[readIdx].value &&
-			boundaryDiff > 0 {
-			// Remove the oldest (readIdx) from the duplicate pair
-			delete(clv.timestamps, clv.changes[readIdx].timestamp)
-			boundaryDiff--
-			changed = true
-			continue
-		}
+	return int(e.kindList[kind].size)
+}
 
-		// If elements have been removed or moved, update the map and the slice
-		if changed {
-			clv.changes[writeIdx] = clv.changes[readIdx]
-		}
+// noEntries checks if there are no entries for the specified `MemberKind`.
+func (e *Entries[T]) noEntries(kind MemberKind) bool {
+	return int(kind) >= len(e.kindList) || e.kindList[kind].size == 0
+}
 
-		writeIdx++
-	}
+//
+// entry, entryList
+//
 
-	if changed {
-		clear(clv.changes[writeIdx:])
-		clv.changes = clv.changes[:writeIdx]
-	}
+// entry is an internal structure representing a single change in the entryList.
+// It includes the kind of the entry, the timestamp of the change, and the value of the change.
+type entry[T comparable] struct {
+	timestamp time.Time // Timestamp of the change.
+	value     T         // Value of the change.
+	prev      *entry[T] // Pointer to the previous entry in the linked list.
+}
+
+// entryList is an internal structure that manages a list of changes (entries) for a specific kind of entry.
+type entryList[T comparable] struct {
+	maxEntries MaxEntries // Maximum number of entries.
+	size       uint8      // Current number of entries.
+	tail       *entry[T]  // Tail pointer for the linked list of entries.
```

Comment:
Why don't we simply use a slice of entry[T] instead?

Reply:
That was my suggestion to Geyslan to potentially overcome the RSS cost. By using linked lists instead of arrays we maintain the same struct layout as before (switching the memberkind for the pointer), but get rid of the slice header. considering the minimal sizes of theharrays we have anyway, linked lists are likely just as good, since we did linear search in the arrays anyway (binary probably would be a very minor imporvement).
In particular the linked lists may be better in terms of memory because we have so little entries per list.

Reply:
That's right. Using node pointers we get rid of 16 extra bytes of slice metadata (len and cap). That was a great idea.

Reply:
I don't think I understand why linked lists will be better with respect to memory.
Using linked list we will be able to remove the prev pointer inside the entry, so an entry will look like this:
```
type entry[T comparable] struct {
	timestamp time.Time // Timestamp of the change.
	value     T         // Value of the change.
}
```
This is 8 bytes per each entry. Since entry is the most inner struct and you can have more than 1 entry in the list, it can easily become 16 or even 24 extra bytes if we consider 3 entries in the list.
In addition to that, we will also be able to remove this extra "size" variable, which is part of the slice metadata. This code will be more elegant, and I think that in most cases won't be less good with respect to memory. I suggest the following struct to be used:
```
type Entries[T comparable] struct {
    entries []struct {
        maxEntries MaxEntries
        entries    []entry[T]
    }
}

type entry[T comparable] struct {
	timestamp time.Time // Timestamp of the change.
	value     T         // Value of the change.
}
```

Reply:
Ok, handling this today. Soon I'll be bringing more data.

Reply:
Ok, after analysis, we opted for slice of slices. Changes pushed again.